### PR TITLE
USB communication threads with ringbuffer

### PR DESCRIPTION
**Pull request by maartentamboer**
USB communication is now working with ringbuffers and a mailbox for the incoming messages. Because protobuf has no framing, a simple framing protocol has been chosen:
`0x55 0xDA <u8_len> [data]` This works great, and can also be used to initially sync the communication in case of serial drivers that send shit over the serial port, just send a couple of `0x55 0xDA 0x00` frames and all comms are in sync.

In the current version the message is sent back, so the protocol stack can be further developed on the PC side

### CODE REVIEW

**Comment by maartentamboer** [comment]
image
Broken build is not my fault 😏

File: include/ringbuffer.h [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+#if (RING_BUFFER_SIZE & (RING_BUFFER_SIZE - 1)) != 0
+#error "RING_BUFFER_SIZE must be a power of two"
+#endif
+
+/**
+ * The type which is used to hold the size
+ * and the indicies of the buffer.
+ * Must be able to fit \c RING_BUFFER_SIZE .
+ */
+typedef uint8_t ring_buffer_size_t;
```
Comment (xor-gate) [inline include/ringbuffer.h]: Don't do this: https://www.kernel.org/doc/html/v4.10/process/coding-style.html#typedefs

File: include/ringbuffer.h [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+/**
+ * Structure which holds a ring buffer.
+ * The buffer contains a buffer array
+ * as well as metadata for the ring buffer.
+ */
+struct ring_buffer_t {
+  /** Buffer memory. */
+  char buffer[RING_BUFFER_SIZE];
+  /** Index of tail. */
+  ring_buffer_size_t tail_index;
```
Comment (xor-gate) [inline include/ringbuffer.h]: For speed use `size_t` as fixed width integers emit assembly code for checking boundaries which is not necessary in this case.

File: include/ringbuffer.h [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+ */
+void ring_buffer_queue_arr(ring_buffer_t *buffer, const char *data, ring_buffer_size_t size);
+
+/**
+ * Returns the oldest byte in a ring buffer.
+ * @param buffer The buffer from which the data should be returned.
+ * @param data A pointer to the location at which the data should be placed.
+ * @return 1 if data was returned; 0 otherwise.
+ */
+uint8_t ring_buffer_dequeue(ring_buffer_t *buffer, char *data);
```
Comment (xor-gate) [inline include/ringbuffer.h]: return value should be more obvious, probably bool. same for the other functions

File: include/ringbuffer.h [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+ */
+uint8_t ring_buffer_peek(ring_buffer_t *buffer, char *data, ring_buffer_size_t index);
+
+
+/**
+ * Returns whether a ring buffer is empty.
+ * @param buffer The buffer for which it should be returned whether it is empty.
+ * @return 1 if empty; 0 otherwise.
+ */
+inline uint8_t ring_buffer_is_empty(ring_buffer_t *buffer) {
```
Comment (xor-gate) [inline include/ringbuffer.h]: return a bool, no need to inline by hand the compilers are smart enough. makes debugging easy. See https://www.kernel.org/doc/html/v4.10/process/coding-style.html#the-inline-disease

File: include/usb_communication.h [hunk 5]
```diff
@@ -0,0 +1,17 @@
+#ifndef USB_COMMUNICATION_H
+#define USB_COMMUNICATION_H
+
+#include <stdint.h>
+
+typedef struct oca_usb_data_frame oca_usb_data_frame;
+struct oca_usb_data_frame
+{
+  char data[128];	//TBD: Depends on the max protobuf frame size
+  uint8_t dataSize;
```
Comment (xor-gate) [inline include/usb_communication.h:10]: size_t

File: src/ringbuffer.c [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+ */
+
+void ring_buffer_init(ring_buffer_t *buffer) {
+  buffer->tail_index = 0;
+  buffer->head_index = 0;
+}
+
+void ring_buffer_queue(ring_buffer_t *buffer, char data) {
+  /* Is buffer full? */
+  if(ring_buffer_is_full(buffer)) {
```
Comment (xor-gate) [inline src/ringbuffer.c:15]: reducing indent level (cyclomatic complexity: https://en.wikipedia.org/wiki/Cyclomatic_complexity) by inverting the statement:

```c
if (!ring_buffer_is_full(buffer))
    return;
buffer->tail_index = ((buffer->tail_index + 1) & RING_BUFFER_MASK);
```

File: src/ringbuffer.c [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+  /* Place data in buffer */
+  buffer->buffer[buffer->head_index] = data;
+  buffer->head_index = ((buffer->head_index + 1) & RING_BUFFER_MASK);
+}
+
+void ring_buffer_queue_arr(ring_buffer_t *buffer, const char *data, ring_buffer_size_t size) {
+  /* Add bytes; one by one */
+  ring_buffer_size_t i;
+  for(i = 0; i < size; i++) {
```
Comment (xor-gate) [inline src/ringbuffer.c]: write like this:

```c
for (size_t n = 0; n < size; n++) {
   ring_buffer_queue(buffer, data[i]);
  // resolve something when queue is full so we know not the full data is written?
}
```

**Review by xor-gate — CHANGES_REQUESTED** [review]
Not completely finished yet, but its a good start for review.

**Comment by maartentamboer** [comment]
Good to see that most of the comments are on the ringbuffer. That's the one I stole from the internet 😛  (https://github.com/AndersKaloer/Ring-Buffer). I will fix the issues soon

File: src/ringbuffer.c (hunk 6 above)
Comment (maartentamboer) [reply src/ringbuffer.c:15]: There are statements after this . So immediately returning doens't seem correct.

```C
void ring_buffer_queue(ring_buffer_t *buffer, char data) {
  /* Is buffer full? */
  if(ring_buffer_is_full(buffer)) {
    /* Is going to overwrite the oldest byte */
    /* Increase tail index */
    buffer->tail_index = ((buffer->tail_index + 1) & RING_BUFFER_MASK);
  }

  /* Place data in buffer */
  buffer->buffer[buffer->head_index] = data;
  buffer->head_index = ((buffer->head_index + 1) & RING_BUFFER_MASK);
}
```

File: include/usb_communication.h (hunk 5 above)
Comment (maartentamboer) [reply include/usb_communication.h:10]: uint8_t because it directly maps to the single size byte in the USB protocol

**Comment by maartentamboer** [comment]
image

Can I merge?

**Pull request merged**